README.md

![Pub Version](https://img.shields.io/pub/v/clean_bottom_navigation_bar?style=for-the-badge) <a href="https://github.com/Solido/awesome-flutter">
   <img alt="Awesome Flutter" src="https://img.shields.io/badge/Awesome-Flutter-blue.svg?longCache=true&style=for-the-badge" />
</a>

# Clean Bottom Navigation Bar

A clean and simple bottom navigation bar with smooth animation and Floating Action Button (FAB) added to the mix.

Check out the source code of the package in the [/lib](https://github.com/0x4d-sh/clean_bottom_navigation_bar/tree/main/lib) folder.

## Preview (default mode)

![Default Mode Gif](screenshots/preview.gif "Clean Flutter Bottom Navigation Bar")

## Getting Started

### Add the plugin:
```yaml
dependencies:
  ...
  clean_bottom_navigation_bar: ^1.0.0
```

### Import the package
```Dart
import 'package:clean_bottom_navigation_bar/clean_bottom_navigation_bar.dart';
```

### Implementation
```Dart
bottomNavigationBar: CleanNavigationBar(
        showFab: true,
        barHeight: 60,
        fabHeight: -30,
        currentIndex: _currentIndex,
        activeColor: Theme.of(context).primaryColor,
        inactiveColor: Theme.of(context).cardColor,
        indicatorColor: Theme.of(context).indicatorColor,
        titleStyle: Theme.of(context)
            .textTheme
            .caption
            .merge(TextStyle(color: Theme.of(context).cardColor)),
        onTap: onTabTapped,
        items: [
          CleanNavigationBarItem(title: Text('Home'), icon: Icons.home),
          CleanNavigationBarItem(title: Text('Search'), icon: Icons.search),
          CleanNavigationBarItem(title: Text('Bag'), icon: Icons.card_travel),
          CleanNavigationBarItem(
              title: Text('Orders'), icon: Icons.shopping_cart),
          CleanNavigationBarItem(
              title: Text('Settings'), icon: Icons.person_outline),
        ],
      ),
```
For more information, please refer to the [example](https://github.com/0x4d-sh/clean_bottom_navigation_bar/tree/main/example) folder.

## Customization
The following variables have been set to allow configuration:

**showFab** - (bool) To show/hide Floating Action Button (FAB)<br />
**fab** - (int) Set which index/tab should the FAB be shown <br />
**cornerRadius** - (double) Set radius to add curves to the corner of the bottom navigation bar<br />
**barHeight** - (double) Specify the bar height of the navigation bar<br />
**fabHeight** - (double) Specify the offset of the FAB to keep inline with the navigation bar<br />
**indicatorHeight** - (double) Set the height for indicator of active tab that is zooming left and right<br />
**activeColor** - (Color) The primary colour for all icons and FAB <br />
**tabColor** - (Color) The background colour of the navigation bar for all tabs <br />
**indicatorColor** - (Color) The color of the indicator that is moving left and right<br />
**titleStyle** - (TextStyle) To specify font, size and color of the text below icons<br />

# Thank you
This package was developed with reference to [Pedromassango](https://github.com/pedromassango/titled_navigation_bar)'s Titled Navigation bar.

CHANGELOG.md
## [1.0.0] - 2020-11-30

* First release of initial clean bottom navigation bar


example/lib/main.dart
import 'package:flutter/material.dart';

import 'navigation.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
          // This is the theme of your application.
          //
          // Try running your application with "flutter run". You'll see the
          // application has a blue toolbar. Then, without quitting the app, try
          // changing the primarySwatch below to Colors.green and then invoke
          // "hot reload" (press "r" in the console where you ran "flutter run",
          // or simply save your changes to "hot reload" in a Flutter IDE).
          // Notice that the counter didn't reset back to zero; the application
          // is not restarted.
          primarySwatch: Colors.blue,
          cardColor: Colors.black,
          indicatorColor: Colors.white),
      home: Navigation(),
    );
  }
}


lib/src/clean_navigation_bar.dart
import 'dart:ui';
import 'package:clean_bottom_navigation_bar/src/clean_navigation_bar_item.dart';
import 'package:flutter/material.dart';

class CleanNavigationBar extends StatefulWidget {
  int currentIndex;
  final bool showFab;
  final int fab;
  final double cornerRadius;
  final double barHeight;
  final double fabHeight;
  final double indicatorHeight;
  final Color activeColor;
  final Color tabColor;
  final Color indicatorColor;
  final TextStyle titleStyle;
  final ValueChanged<int> onTap;
  final List<CleanNavigationBarItem> items;

  CleanNavigationBar({
    Key key,
    this.showFab = true,
    this.fab = 2,
    this.currentIndex = 0,
    this.cornerRadius = 20,
    this.barHeight = 60,
    this.fabHeight = 0 - (60 / 2),
    this.indicatorHeight = 4,
    this.activeColor,
    this.tabColor,
    this.indicatorColor,
    this.titleStyle,
    @required this.onTap,
    @required this.items,
  }) : super(key: key);

  @override
  _NeatNavigationBarState createState() => _NeatNavigationBarState();
}

class _NeatNavigationBarState extends State<CleanNavigationBar> {
  double get BAR_HEIGHT => widget.barHeight;
  double get INDICATOR_HEIGHT => widget.indicatorHeight;
  List<CleanNavigationBarItem> get items => widget.items;

  double BAR_ITEM_WIDTH = 0;
  Duration duration = Duration(milliseconds: 150);

  @override
  Widget build(BuildContext context) {
    double BAR_ITEM_WIDTH = MediaQuery.of(context).size.width / items.length;
    return Container(
      height: BAR_HEIGHT + MediaQuery.of(context).viewPadding.bottom,
      width: MediaQuery.of(context).size.width,
      decoration: BoxDecoration(
        color: widget.activeColor,
        borderRadius: BorderRadius.only(
          topLeft: Radius.circular(widget.cornerRadius),
          topRight: Radius.circular(widget.cornerRadius),
        ),
      ),
      child: Stack(
        clipBehavior: Clip.none,
        children: <Widget>[
          Positioned(
            top: INDICATOR_HEIGHT,
            child: Row(
              crossAxisAlignment: CrossAxisAlignment.end,
              children: items.map((item) {
                var index = items.indexOf(item);
                return GestureDetector(
                    onTap: () => _select(index),
                    child: _buildItem(
                        item,
                        index == widget.currentIndex,
                        (index == widget.fab)
                            ? (widget.showFab)
                                ? true
                                : false
                            : false));
              }).toList(),
            ),
          ),
          Positioned(
            top: 0,
            width: MediaQuery.of(context).size.width,
            child: AnimatedAlign(
              alignment:
                  Alignment(_getIndicatorPosition(widget.currentIndex), 0),
              curve: Curves.linear,
              duration: duration,
              child: Container(
                width: BAR_ITEM_WIDTH,
                height: INDICATOR_HEIGHT,
                child: Row(
                  children: <Widget>[
                    Container(
                      width: (BAR_ITEM_WIDTH / 4),
                      color: Color(0x00000000),
                    ),
                    Container(
                      width: (BAR_ITEM_WIDTH / 2),
                      decoration: BoxDecoration(
                        color: widget.indicatorColor,
                        borderRadius: BorderRadius.circular(10),
                      ),
                    ),
                    Container(
                      width: (BAR_ITEM_WIDTH / 4),
                      color: Color(0x00000000),
                    ),
                  ],
                ),
              ),
            ),
          ),
          if (widget.showFab)
            Positioned(
              top: widget.fabHeight,
              child: Row(
                crossAxisAlignment: CrossAxisAlignment.end,
                children: items.map((item) {
                  var index = items.indexOf(item);
                  return (index == widget.fab)
                      ? GestureDetector(
                          onTap: () => _select(widget.fab),
                          child: _buildButton(item, index == widget.fab))
                      : Container(
                          width:
                              MediaQuery.of(context).size.width / items.length);
                }).toList(),
              ),
            ),
        ],
      ),
    );
  }

  _select(int index) {
    widget.currentIndex = index;
    widget.onTap(widget.currentIndex);

    setState(() {});
  }

  double _getIndicatorPosition(int index) {
    var isLtr = Directionality.of(context) == TextDirection.ltr;
    if (isLtr)
      return lerpDouble(-1.0, 1.0, index / (items.length - 1));
    else
      return lerpDouble(1.0, -1.0, index / (items.length - 1));
  }

  Widget _buildIcon(CleanNavigationBarItem item, bool active, bool fab) {
    return Icon(
      item.icon,
      color: fab ? widget.activeColor : widget.tabColor,
    );
  }

  Widget _buildText(CleanNavigationBarItem item, bool active) {
    return DefaultTextStyle.merge(
      child: item.title,
      style: widget.titleStyle,
    );
  }

  Widget _buildItem(CleanNavigationBarItem item, bool active, bool fab) {
    return Container(
      height: BAR_HEIGHT,
      width: MediaQuery.of(context).size.width / items.length,
      child: Column(
        children: <Widget>[
          Container(
            padding: EdgeInsets.only(top: 10.0, bottom: 5.0),
            child: _buildIcon(item, active, fab),
          ),
          _buildText(item, active),
        ],
      ),
    );
  }

  Widget _buildButton(CleanNavigationBarItem item, bool fab) {
    return Container(
      width: MediaQuery.of(context).size.width / items.length,
      child: AnimatedContainer(
        child: _buildIcon(item, fab, true),
        decoration: BoxDecoration(
          shape: BoxShape.circle,
          color: (widget.currentIndex == widget.fab)
              ? widget.indicatorColor
              : widget.tabColor,
        ),
        width: (MediaQuery.of(context).size.width / items.length) * (70 / 100),
        height: (MediaQuery.of(context).size.width / items.length) * (70 / 100),
        duration: duration,
      ),
    );
  }
}


example/lib/navigation.dart
import 'package:clean_bottom_navigation_bar/clean_bottom_navigation_bar.dart';
import 'package:flutter/material.dart';

import 'blank.dart';

class Navigation extends StatefulWidget {
  @override
  _NavigationState createState() => _NavigationState();
}

class _NavigationState extends State<Navigation> {
  int _currentIndex = 0;
  final List<Widget> _children = [
    Blank(),
    Blank(),
    Blank(),
    Blank(),
    Blank(),
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      bottomNavigationBar: CleanNavigationBar(
        showFab: true,
        barHeight: 60,
        fabHeight: -30,
        currentIndex: _currentIndex,
        activeColor: Theme.of(context).primaryColor,
        tabColor: Theme.of(context).cardColor,
        indicatorColor: Theme.of(context).indicatorColor,
        titleStyle: Theme.of(context)
            .textTheme
            .caption
            .merge(TextStyle(color: Theme.of(context).cardColor)),
        onTap: onTabTapped,
        items: [
          CleanNavigationBarItem(title: Text('Home'), icon: Icons.home),
          CleanNavigationBarItem(title: Text('Search'), icon: Icons.search),
          CleanNavigationBarItem(title: Text('Bag'), icon: Icons.card_travel),
          CleanNavigationBarItem(
              title: Text('Orders'), icon: Icons.shopping_cart),
          CleanNavigationBarItem(
              title: Text('Settings'), icon: Icons.person_outline),
        ],
      ),
      body: _children[_currentIndex],
      backgroundColor: Theme.of(context).backgroundColor,
    );
  }

  void onTabTapped(int index) {
    setState(() {
      _currentIndex = index;
    });
  }
}
